### ISSUE TITLE
Issues with the text element function API

### ISSUE THREAD

**Issue by 9SMTM6** [typst/typst#1795]
### Description

While considering previous issues of mine (#1783, #1784) I looked into the implementation of and experimented with the text function.

I've found a number of flaws, not all of them are easy to trace in the source code, and not all of them may end up being accepted, but some of them are IMO very crucial, and the solutions might also influence each other.

I've decided against a PR, other than what I said earlier as some things are more complex, and more importantly the solution ought to be discussed.

So first of all, lets start by listing the identified issues (as far as I can tell these are real, though If they have a solution I missed I'd love to hear):

* a number of named settings are not really toggles and only apply with specific values - eg `ligatures`only applies when true, `kerning` only when false, if the inverse of these values is "set" then they are just ignored and the font default (?) applies (see https://github.com/typst/typst/blob/f197a447a4aea5f87e9a5a5f04a59fe5b36a82b0/crates/typst-library/src/text/shaping.rs#L877).
  * with some settings there were some efforts to document this, but they are easy to miss, its not documented in all instances, and frankly that doing `#set text(ligatures=false)` and that doing nothing is just very unintuitive. 
    * If this behavior should be kept, they should be renamed, eg to `force_ligatures`
    * If we go away from this behavior boolean values don't suffice, otherwise theres no way to "take back" these settings and go to the font default. We should probably use string tags, if possible, such as `force_enable`, `force_disable`, `font_default`
 * because of the behavior of the style chain the ligature settings survive the setting of different fonts, which may also be unexpected for newcomers. This makes it more important to have a way to switch back to the default behavior.
 * applying raw OpenType features (`features` dict) ignores subsequent settings of the same feature entirely. SOMETHING leads to problematic ordering of these features? I'm not sure how the features get applied in the end, all of this is macro heavy and thus difficult to trace in the source code, as I can't just search for usages (see #1784 comments after closing for examples).
 * raw features also influence the other settings, though that is probably intended and should remain, listing it for completeness
 * raw features have no way to defer to the font default after having them set once.

I've already given some suggestions for solutions while listing issues.
I think it should also be possible to reset these settings, and this should perhaps actually be the default here. If I use `#set text(font="blahblah")` I kind of expect to get the default behavior of the font.

For libraries and usable functions that use the context there probably also should be a way to keep selected parts of the settings so far, and only selectively ensure some OpenType setting is set a certain way.

I'm unclear how these later issues should be solved in the end.

However the behavior where the first setting of a raw OpenType Feature wins is an outright bug and has to be solved, and I think I've made some other suggestions (such as `force_*` and `default` instead of booleans) that would be a clear improvement in making the API easier to understand.

### Typst version

- [X] I am using the latest version of Typst

**Comment by PgBiel** [comment]
Small comment:
> (such as `force_*` and `default` instead of booleans)

I believe what you're looking for is `Smart<bool>`, a.k.a. `auto` for the default and `true/false` for an override.

**Comment by laurmaedje** [comment]
I just skimmed quickly and will come back to this issue later (if I don't in ~1 week feel free to remind me), but wanted to remark one thing: It's very possible that it's broken, but the intent of `!TextElem::kerning_in(styles)` was not to select the _font's default,_ but the _shaping engine's default._ Rustybuzz/HarfBuzz internally enable some features by default and we skip setting OpenType features if they align with rustybuzz's defaults.

**Comment by 9SMTM6** [comment]
> I believe what you're looking for is `Smart<bool>`, a.k.a. `auto` for the default and `true/false` for an override.

I guess. I like the expressiveness of strings, but they are doubtlessly more cumbersome to use and the style you propose is probably more in line with other APIs.



> was not to select the font's default, but the shaping engine's default. Rustybuzz/HarfBuzz internally enable some features by default

I should probably note that I'm all but an expert on OpenType Fonts, features, or Harfbuzz. I've heard of it in the context of the tectonic project, but that's all. I noted something to that extend in one of my earlier issues, but forgot to reiterate it here.

I don't even know if fonts CAN have defaults. For all I know my impression, that they have, could've just been a result of ligatures being behind different features, that Harf/Rustbuzz had different defaults for.

If fonts don't have defaults then this point

 > * a number of named settings are not really toggles and only apply with specific values - eg ligaturesonly applies when true, kerning only when false, if the inverse of these values is "set" then they are just ignored and the font default (?) applies (see 

(and others) is wrong. I'd still want a way to go back to the "default" behavior without consulting some unrelated documentation  though, so using the flags as I proposed or Smart<bool> would IMO be an improvement.

When it comes to doing what you're doing there too, why is this done? Does this make some kind of performance difference? I don't get entirely get this choice. It seems to me as If one is asking for unintended behavior. AFAICT it works out here, because of the behavior of `[attribute]_in` only selecting the last one, but if that assumption is broken somewhere /sometime later it could cause issues.

**Comment by 9SMTM6** [comment]
> * applying raw OpenType features (features dict) ignores subsequent settings of the same feature entirely. SOMETHING leads to problematic ordering of these features? I'm not sure how the features get applied in the end, all of this is macro heavy and thus difficult to trace in the source code, as I can't just search for usages (see https://github.com/typst/typst/issues/1784 comments after closing for examples).

This point is easy to misunderstand. I thought when going through this right now that I made a mistake, only after a re-read its correct.

Trying to formulate it in a clearer way: If you mention a raw feature once, its set in stone. But if you have a subsequent `#set text(features(...)` then NEW features mentioned there still apply. 

I've found a resolution for this particular issue. Reversing `features_in(self).0` (a Vector) in the following snippet solves this:

https://github.com/typst/typst/blob/66df130ca4c9217d137cba0778d33c8ff9349d33/crates/typst-library/src/text/shaping.rs#L929-L931

took a bit to wrap my head around that, but it makes sense now. The last features in that `tags` Vector win.

There is another issue with the API btw., while we're in that behavior, though its really just a worse case of an issue mentioned earlier: If we set a raw feature that has a corresponding option of `text`, then any subsequent setting of this corresponding option will always be ignored. So this:

```typ
#set text(features: (dlig: 1))
#set text(discretionary-ligatures: false)
```
means that subsequently discretionary-ligatures is still enabled.

This is because the ordering of the different options of the element function in the compiler source wins over the ordering of the calls:

https://github.com/typst/typst/blob/66df130ca4c9217d137cba0778d33c8ff9349d33/crates/typst-library/src/text/shaping.rs#L901C1-L903C6

this is above the application of the features option, and ordering in the text file is only evaluated per-option. Thus features wins always.
Its another argument to reset the features by default if using `text`. If unchanged then at the very least this behavior should be documented.

**Comment by 9SMTM6** [comment]
BTW, I'm going on vacation soon, tuesday morning German time. Its for a whole month (finished bachelor, starting master only in Oct). So I'll likely not be very responsive.

Sorry for that timing, perhaps I should not have opened such a big issue with that timing.

**Comment by laurmaedje** [comment]
> Sorry for that timing, perhaps I should not have opened such a big issue with that timing.

No problem! Then, I have a bit more time to think through the problem.
